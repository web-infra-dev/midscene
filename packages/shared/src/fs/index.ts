import assert from 'node:assert';
import { existsSync, readFileSync } from 'node:fs';
import { dirname, join } from 'node:path';

interface PkgInfo {
  name: string;
  version: string;
  dir: string;
}

let pkg: PkgInfo | undefined;
export function getMidscenePkgInfo(dir: string): PkgInfo {
  if (pkg) {
    return pkg;
  }

  const pkgDir = findNearestPackageJson(dir || process.cwd());
  assert(pkgDir, 'package.json not found');
  const pkgJsonFile = join(pkgDir, 'package.json');

  if (pkgJsonFile) {
    const { name, version } = JSON.parse(readFileSync(pkgJsonFile, 'utf-8'));
    pkg = { name, version, dir: pkgDir };
    return pkg;
  }
  return {
    name: 'midscene-unknown-page-name',
    version: '0.0.0',
    dir: pkgDir,
  };
}

/**
 * Find the nearest package.json file recursively
 * @param {string} dir - Home directory
 * @returns {string|null} - The most recent package.json file path or null
 */
export function findNearestPackageJson(dir: string): string | null {
  const packageJsonPath = join(dir, 'package.json');
  if (existsSync(packageJsonPath)) {
    return dir;
  }

  const parentDir = dirname(dir);

  // Return null if the root directory has been reached
  if (parentDir === dir) {
    return null;
  }

  return findNearestPackageJson(parentDir);
}
