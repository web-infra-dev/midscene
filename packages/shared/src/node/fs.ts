import { existsSync, readFileSync } from 'node:fs';
import { dirname, join } from 'node:path';
import { ifInBrowser, ifInWorker } from '../utils';

declare const __HTML_ELEMENT_SCRIPT__: string;

interface PkgInfo {
  name: string;
  version: string;
  dir: string;
}

const pkgCacheMap: Record<string, PkgInfo> = {};

export function getRunningPkgInfo(dir?: string): PkgInfo | null {
  if (ifInBrowser || ifInWorker) {
    return null;
  }
  const dirToCheck = dir || process.cwd();
  if (pkgCacheMap[dirToCheck]) {
    return pkgCacheMap[dirToCheck];
  }

  const pkgDir = findNearestPackageJson(dirToCheck);
  const pkgJsonFile = pkgDir ? join(pkgDir, 'package.json') : null;

  if (pkgDir && pkgJsonFile) {
    const { name, version } = JSON.parse(readFileSync(pkgJsonFile, 'utf-8'));
    pkgCacheMap[dirToCheck] = {
      name: name || 'midscene-unknown-package-name',
      version: version || '0.0.0',
      dir: pkgDir,
    };
    return pkgCacheMap[dirToCheck];
  }
  return {
    name: 'midscene-unknown-package-name',
    version: '0.0.0',
    dir: dirToCheck,
  };
}

/**
 * Find the nearest package.json file recursively
 * @param {string} dir - Home directory
 * @returns {string|null} - The most recent package.json file path or null
 */
export function findNearestPackageJson(dir: string): string | null {
  const packageJsonPath = join(dir, 'package.json');
  if (existsSync(packageJsonPath)) {
    return dir;
  }

  const parentDir = dirname(dir);

  // Return null if the root directory has been reached
  if (parentDir === dir) {
    return null;
  }

  return findNearestPackageJson(parentDir);
}

export function getElementInfosScriptContent() {
  const htmlElementScript = __HTML_ELEMENT_SCRIPT__;

  if (!htmlElementScript) {
    throw new Error('HTML_ELEMENT_SCRIPT inject failed.');
  }
  return htmlElementScript;
}

export async function getExtraReturnLogic(tree = false) {
  if (ifInBrowser || ifInWorker) {
    return null;
  }

  const elementInfosScriptContent = `${getElementInfosScriptContent()}midscene_element_inspector.setNodeHashCacheListOnWindow();`;

  if (tree) {
    return `${elementInfosScriptContent}midscene_element_inspector.webExtractNodeTree()`;
  }
  return `${elementInfosScriptContent}midscene_element_inspector.webExtractTextWithPosition()`;
}
