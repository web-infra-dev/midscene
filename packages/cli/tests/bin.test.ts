import { randomUUID } from 'node:crypto';
import { existsSync, unlinkSync } from 'node:fs';
import { execa } from 'execa';
import { describe, expect, test, vi } from 'vitest';

const cliBin = require.resolve('../bin/midscene');
vi.setConfig({
  testTimeout: 30 * 1000,
});
describe.skipIf(process.platform !== 'darwin')('bin', () => {
  test('error order', async () => {
    const params = [
      '--query',
      '{name: string, status: string}[], service status of github page',
      '--url',
      'https://www.baidu.com/',
    ];
    expect(async () => {
      await execa(cliBin, params);
    }).rejects.toThrowError();
  });

  test('basic action', async () => {
    const randomFileName = `status-${randomUUID()}.json`;
    const params = [
      '--url',
      'https://www.githubstatus.com/',
      '--query-output',
      randomFileName,
      '--query',
      '{name: string, status: string}[], service status of github page',
    ];
    const { failed } = await execa(cliBin, params);
    expect(failed).toBe(false);

    expect(existsSync(randomFileName)).toBeTruthy();
    unlinkSync(randomFileName);
  });
});
