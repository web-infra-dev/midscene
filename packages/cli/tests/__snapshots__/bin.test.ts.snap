// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`bin > local server - evaluateJavaScript 1`] = `
{
  "foo": "bar",
  "promise": "hello",
}
`;
